### PR TITLE
incremental: support postbuild phase

### PR DESCRIPTION
**Pull request by ansyral**
@vwxyzh @vicancy @superyyrrzz @hellosnow @qinezh @chenkennt

### CODE REVIEW

File: src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs [hunk 1]
```diff
@@ -454,6 +454,15 @@ private bool GetCanVersionIncremental()
                 Logger.LogVerbose($"Disable incremental build by force rebuild option.");
                 return false;
             }
+            if (_parameters.ApplyTemplateSettings != null)
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs]: >ApplyTemplateSettings [](start = 28, length = 21)

it's create by internal and won't be null?

**Comment by hellosnow** [comment]
:shipit:

File: src/Microsoft.DocAsCode.Build.Engine/DocumentBuildParameters.cs [hunk 2]
```diff
@@ -31,8 +31,8 @@ public sealed class DocumentBuildParameters : MarshalByRefObject
 
         public TemplateManager TemplateManager { get; set; }
 
-        // todo : partial properties should check.
-        //[IncrementalCheck]
+        // todo : move the check to the prehandle of postbuild phase
+        [IncrementalCheck]
         public ApplyTemplateSettings ApplyTemplateSettings { get; set; }
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.Engine/DocumentBuildParameters.cs]: `ApplyTemplateSettings` contains absolute path, that's why I remove the attribute `IncrementalCheck`

File: src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs (hunk 1 above)
Comment (ansyral) [inline src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs]: checked in Create Method

File: src/Microsoft.DocAsCode.Build.Engine/DocumentBuildParameters.cs (hunk 2 above)
Comment (ansyral) [inline src/Microsoft.DocAsCode.Build.Engine/DocumentBuildParameters.cs]: i should use IncrementalIgnore Attribute as the customized ContractResolver would check the property recursively

File: src/Microsoft.DocAsCode.Build.Engine/DocumentBuildParameters.cs [hunk 3]
```diff
@@ -10,47 +10,47 @@ namespace Microsoft.DocAsCode.Build.Engine
...
 
         public string RootTocPath { get; set; }
 
+        [IncrementalIgnore]
         public string TemplateDir { get; set; }
 
+        [IncrementalIgnore]
         public ImmutableDictionary<string, ChangeKindWithDependency> Changes { get; set; }
 
+        [IncrementalIgnore]
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/DocumentBuildParameters.cs]: >IncrementalIgnore [](start = 9, length = 17)

when force rebuild, should we always full build? What happens when IncrementalIgnore this property?

Comment (ansyral) [inline src/Microsoft.DocAsCode.Build.Engine/DocumentBuildParameters.cs]: when force rebuild, we'll always full build. the attribute IncrementalIgnore means that the property/field would be ignored when calculating config hash

---
In reply to: [90815153](https://github.com/dotnet/docfx/pull/953#discussion_r90815153) [](ancestors = 90815153)

**Pull request merged**